metadata/content.go

```diff
@@ -19,16 +21,25 @@ import (
 
 type contentStore struct {
 	content.Store
-	db *bolt.DB
+	db *DB
+
+	l sync.RWMutex
 }
 
 // NewContentStore returns a namespaced content store using an existing
 // content store interface.
-func NewContentStore(db *bolt.DB, cs content.Store) content.Store {
-	return &contentStore{
-		Store: cs,
-		db:    db,
+func NewContentStore(db *DB, cs content.Store) content.Store {
+	db.storeL.Lock()
+	defer db.storeL.Unlock()
+
+	if db.cs == nil {
```

Comment:
This is a little odd. We shouldn't rely on this fixup on the database. Are we not able to construct the database with the correct content store?

Reply:
I was thinking that we would want to prevent overriding, however it doesn't really make sense for it to be called multiple times. I was trying to prevent changes to other parts of the code that handle this initialization, it is already somewhat complicated by the fact that the metadata db is passed around as a plugin and these functions are also initialized by plugins. The alternative to have the database object fully initialized at construction would require the metadata plugin to rely on content store and snapshot store and the services to only rely on the metadata store. That pattern may make more sense now.

Reply:
I think it is important that we ensure our initialization is worked out correctly. Usually, ergonomics problems during instantiation reflect a problem with the object model. We might be missing a component.

Let's draw a dependency diagram and work it out from there.